test/test_torch.py

```diff
@@ -11001,6 +11001,69 @@ def test_exponential(self, device, dtype):
         with self.assertRaises(RuntimeError):
             torch.empty((1,), device=device, dtype=dtype).exponential_(-0.5)
 
+    @unittest.skipIf(not TEST_NUMPY, "Numpy not found")
+    @dtypes(*(torch.testing.get_all_fp_dtypes(include_half=False) +
+              torch.testing.get_all_complex_dtypes()))
+    @dtypesIfCUDA(*(torch.testing.get_all_fp_dtypes(include_half=True) +
+                    torch.testing.get_all_complex_dtypes()))
+    def test_exp(self, device, dtype):
+        for v in (2, -2) + ((1j, 1 + 1j) if dtype.is_complex else ()):
+            if dtype == torch.bfloat16:
+                # Currently multiply a bfloat16 type with floating-point causes error:
+                #   RuntimeError: dtype != ScalarType::Undefined INTERNAL ASSERT FAILED at
+                #   "/pytorch/aten/src/ATen/native/TensorIterator.cpp":125, please report a bug to PyTorch.
+                # We skip bfloat16 for now, but we should fix it.
+                if not device.startswith('xla'):
+                    with self.assertRaises(RuntimeError):
+                        torch.tensor(v, dtype=dtype, device=device) * torch.arange(18, device=device)
+                    return
+
+            a = torch.tensor(v, dtype=dtype, device=device) * torch.arange(18, device=device) / 3 * math.pi
+            expected = np.exp(a.cpu().numpy())
+            actual = torch.exp(a)
+            self.assertEqual(actual, torch.from_numpy(expected))
+
+            if dtype.is_complex:
+                inf_real_zero_imag = torch.tensor(complex(float('inf'), 0), device=device, dtype=dtype)
+                if device.startswith('cpu'):
+                    # This is incorrect. It should be zero. Need fix!
+                    self.assertNotEqual(torch.exp(inf_real_zero_imag).item().imag, 0)
+                else:
```

Comment:
I think we should still compare with the numpy result for all the special cases as well

Reply:
The checks here are consistent with NumPy results and is explicit in what NumPy exp would lead to.

Reply:
Yes but sometimes NumPy changes and it's important we detect those changes, too. For example, NumPy changed how they convert float to int numbers in some cases when running on our MacOS build.

Reply:
OK, I added numpy tests.